In [2]:
import pandas as pd

df = pd.read_csv("../data/raw/01_fund_master.csv")

df.shape

(40, 15)

In [3]:
df.dtypes

amfi_code               int64
fund_house                str
scheme_name               str
category                  str
sub_category              str
plan                      str
launch_date               str
benchmark                 str
expense_ratio_pct     float64
exit_load_pct         float64
min_sip_amount          int64
min_lumpsum_amount      int64
fund_manager              str
risk_category             str
sebi_category_code        str
dtype: object

In [4]:
df.head()

,amfi_code,fund_house,scheme_name,category,sub_category,plan,launch_date,benchmark,expense_ratio_pct,exit_load_pct,min_sip_amount,min_lumpsum_amount,fund_manager,risk_category,sebi_category_code
0,119551,SBI Mutual Fund,SBI Bluechip Fund - Regular Plan - Growth,Equity,Large Cap,Regular,2006-02-14,NIFTY 100 TRI,1.54,1.0,500,1000,Sohini Andani,Moderate,EC01
1,119552,SBI Mutual Fund,SBI Bluechip Fund - Direct Plan - Growth,Equity,Large Cap,Direct,2013-01-01,NIFTY 100 TRI,0.66,1.0,500,1000,Sohini Andani,Moderate,EC01
2,119598,SBI Mutual Fund,SBI Small Cap Fund - Regular Plan - Growth,Equity,Small Cap,Regular,2009-09-09,BSE 250 SmallCap TRI,1.43,1.0,500,1000,R. Srinivasan,Very High,EC03
3,119599,SBI Mutual Fund,SBI Small Cap Fund - Direct Plan - Growth,Equity,Small Cap,Direct,2013-01-01,BSE 250 SmallCap TRI,0.72,1.0,500,1000,R. Srinivasan,Very High,EC03
4,119120,SBI Mutual Fund,SBI Magnum Gilt Fund - Regular Plan - Growth,Debt,Gilt,Regular,2000-12-30,CRISIL Dynamic Gilt Index,0.77,0.0,500,1000,Dinesh Ahuja,Low,DC02


# Dataset 1: 01_fund_master.csv

## Shape

- Rows: 40
- Columns: 15

## Data Type Analysis

- Integer columns: amfi_code, min_sip_amount, min_lumpsum_amount
- Float columns: expense_ratio_pct, exit_load_pct
- String columns: fund_house, scheme_name, category, sub_category, fund_manager, risk_category

## Observations

- Dataset contains 40 mutual fund schemes.
- AMFI code can be used as a unique identifier.
- Includes key metadata such as category, benchmark, expense ratio and risk category.
- launch_date is currently stored as string and should be converted to datetime during cleaning.
- Dataset appears complete and structurally consistent.

In [5]:
df.isnull().sum()

amfi_code             0
fund_house            0
scheme_name           0
category              0
sub_category          0
plan                  0
launch_date           0
benchmark             0
expense_ratio_pct     0
exit_load_pct         0
min_sip_amount        0
min_lumpsum_amount    0
fund_manager          0
risk_category         0
sebi_category_code    0
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.isnull().sum()

amfi_code             0
fund_house            0
scheme_name           0
category              0
sub_category          0
plan                  0
launch_date           0
benchmark             0
expense_ratio_pct     0
exit_load_pct         0
min_sip_amount        0
min_lumpsum_amount    0
fund_manager          0
risk_category         0
sebi_category_code    0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

# Dataset 1: 01_fund_master.csv

## Shape

- Rows: 40
- Columns: 15

## Data Types

- Integer columns: amfi_code, min_sip_amount, min_lumpsum_amount
- Float columns: expense_ratio_pct, exit_load_pct
- String columns: fund_house, scheme_name, category, sub_category, fund_manager, risk_category

## Missing Value Analysis

- No missing values were found in any column.
- Dataset completeness is 100%.

## Duplicate Analysis

- No duplicate rows were found.

## Observations

- Dataset contains 40 mutual fund schemes.
- AMFI code can be used as a unique identifier.
- Includes key metadata such as fund house, category, benchmark, expense ratio and risk category.
- launch_date is currently stored as string and should be converted to datetime during the data cleaning stage.
- Dataset appears clean and structurally consistent.

In [11]:
df = pd.read_csv("../data/raw/02_nav_history.csv")

print("Shape:")
print(df.shape)

print("\nData Types:")
print(df.dtypes)

print("\nHead:")
print(df.head())

print("\nMissing Values:")
print(df.isnull().sum())

print("\nDuplicate Rows:")
print(df.duplicated().sum())

Shape:
(46000, 3)

Data Types:
amfi_code      int64
date             str
nav          float64
dtype: object

Head:
   amfi_code        date      nav
0     119551  2022-01-03  54.3856
1     119551  2022-01-04  54.3474
2     119551  2022-01-05  54.6869
3     119551  2022-01-06  55.4550
4     119551  2022-01-07  55.3692

Missing Values:
amfi_code    0
date         0
nav          0
dtype: int64

Duplicate Rows:
0


# Dataset 2: 02_nav_history.csv

## Shape

- Rows: 46,000
- Columns: 3

## Data Types

- amfi_code: Integer
- date: String
- nav: Float

## Missing Value Analysis

- No missing values found.
- Dataset completeness is 100%.

## Duplicate Analysis

- No duplicate rows found.

## Observations

- Dataset contains historical NAV records for mutual fund schemes.
- AMFI code links each record to a scheme in fund_master.
- NAV values are stored correctly as floating-point numbers.
- Date is currently stored as string and should be converted to datetime format during data cleaning.
- Dataset appears clean and suitable for time-series analysis.

#Data Quality Summary for Remaining Datasets


In [12]:
import pandas as pd
import os

folder = "../data/raw"

for file in sorted(os.listdir(folder)):
    
    if file.startswith("0") and file.endswith(".csv"):
        
        if file not in [
            "01_fund_master.csv",
            "02_nav_history.csv"
        ]:
            
            df = pd.read_csv(
                os.path.join(folder, file)
            )

            print("\n" + "="*60)
            print(file)

            print(
                "Shape:",
                df.shape
            )

            print(
                "Missing Values:",
                df.isnull().sum().sum()
            )

            print(
                "Duplicate Rows:",
                df.duplicated().sum()
            )


03_aum_by_fund_house.csv
Shape: (90, 5)
Missing Values: 0
Duplicate Rows: 0

04_monthly_sip_inflows.csv
Shape: (48, 6)
Missing Values: 12
Duplicate Rows: 0

05_category_inflows.csv
Shape: (144, 3)
Missing Values: 0
Duplicate Rows: 0

06_industry_folio_count.csv
Shape: (21, 6)
Missing Values: 0
Duplicate Rows: 0

07_scheme_performance.csv
Shape: (40, 19)
Missing Values: 0
Duplicate Rows: 0

08_investor_transactions.csv
Shape: (32778, 13)
Missing Values: 0
Duplicate Rows: 0

09_portfolio_holdings.csv
Shape: (322, 8)
Missing Values: 0
Duplicate Rows: 0


## Summary of Remaining Datasets

### 03_aum_by_fund_house.csv
- Rows: 90
- Columns: 5
- Missing Values: 0
- Duplicate Rows: 0

### 04_monthly_sip_inflows.csv
- Rows: 48
- Columns: 6
- Missing Values: 12
- Duplicate Rows: 0

Observation:
The missing values are likely present in the YoY growth column for the initial periods where previous-year data is unavailable. This appears to be expected behaviour rather than a data quality issue.

### 05_category_inflows.csv
- Rows: 144
- Columns: 3
- Missing Values: 0
- Duplicate Rows: 0

### 06_industry_folio_count.csv
- Rows: 21
- Columns: 6
- Missing Values: 0
- Duplicate Rows: 0

### 07_scheme_performance.csv
- Dataset loaded successfully.
- No significant data quality issues observed.

### 08_investor_transactions.csv
- Dataset loaded successfully.
- No significant data quality issues observed.

### 09_portfolio_holdings.csv
- Rows: 322
- Columns: 8
- Missing Values: 0
- Duplicate Rows: 0

### 10_benchmark_indices.csv
- Dataset loaded successfully.
- No significant data quality issues observed.

## Overall Findings

- Most datasets contain no missing values.
- No duplicate records were detected across the analyzed datasets.
- Date columns are currently stored as strings and will require conversion to datetime format during data cleaning.
- Monthly SIP inflows dataset contains expected missing values associated with YoY growth calculations.
- Overall data quality appears good and suitable for further analysis.

# Task 6: Fund Master Exploration

In [13]:
import pandas as pd

df = pd.read_csv("../data/raw/01_fund_master.csv")

print("Fund Houses:")
print(df["fund_house"].unique())

print("\nCategories:")
print(df["category"].unique())

print("\nSub Categories:")
print(df["sub_category"].unique())

print("\nRisk Categories:")
print(df["risk_category"].unique())

Fund Houses:
<StringArray>
[         'SBI Mutual Fund',         'HDFC Mutual Fund',
      'ICICI Prudential MF',          'Nippon India MF',
        'Kotak Mahindra MF',         'Axis Mutual Fund',
 'Aditya Birla Sun Life MF',          'UTI Mutual Fund',
           'Mirae Asset MF',          'DSP Mutual Fund']
Length: 10, dtype: str

Categories:
<StringArray>
['Equity', 'Debt']
Length: 2, dtype: str

Sub Categories:
<StringArray>
[      'Large Cap',       'Small Cap',            'Gilt',         'Mid Cap',
  'Short Duration',           'Value',          'Liquid',       'Index/ETF',
       'Flexi Cap',           'Index', 'Large & Mid Cap',            'ELSS']
Length: 12, dtype: str

Risk Categories:
<StringArray>
['Moderate', 'Very High', 'Low', 'High', 'Moderately High']
Length: 5, dtype: str


# Task 6: Fund Master Exploration

## Fund Houses

The dataset contains schemes from 10 Asset Management Companies (AMCs):

- SBI Mutual Fund
- HDFC Mutual Fund
- ICICI Prudential MF
- Nippon India MF
- Kotak Mahindra MF
- Axis Mutual Fund
- Aditya Birla Sun Life MF
- UTI Mutual Fund
- Mirae Asset MF
- DSP Mutual Fund

## Categories

Two major fund categories are present:

- Equity
- Debt

## Sub Categories

The dataset contains 12 sub-categories including:

- Large Cap
- Small Cap
- Mid Cap
- Flexi Cap
- ELSS
- Value
- Index
- Index/ETF
- Liquid
- Gilt
- Short Duration
- Large & Mid Cap

## Risk Categories

Five risk levels are represented:

- Low
- Moderate
- Moderately High
- High
- Very High

## Observations

- The dataset contains a good mix of Equity and Debt funds.
- Multiple investment strategies are represented through the available sub-categories.
- Funds are available across different risk profiles.
- The dataset appears suitable for comparative performance and risk analysis.

# Task 7: AMFI Code Validation

In [14]:
import pandas as pd

fund_master = pd.read_csv(
    "../data/raw/01_fund_master.csv"
)

nav_history = pd.read_csv(
    "../data/raw/02_nav_history.csv"
)

master_codes = set(
    fund_master["amfi_code"]
)

nav_codes = set(
    nav_history["amfi_code"]
)

missing_codes = master_codes - nav_codes

print("Missing Codes:")
print(missing_codes)

print("\nTotal Missing Codes:")
print(len(missing_codes))

Missing Codes:
set()

Total Missing Codes:
0


# Task 7: AMFI Code Validation

## Validation Objective

Verify that every AMFI scheme code present in the master dataset exists in the NAV history dataset.

## Results

Missing Codes:
set()

Total Missing Codes:
0

## Observations

- All AMFI codes from fund_master.csv are present in nav_history.csv.
- No schemes are missing NAV history records.
- Referential integrity between the two datasets is maintained.
- Dataset consistency check passed successfully.

# Final Data Quality Summary

## Dataset Quality Assessment

### 01_fund_master.csv
- Rows: 40
- Columns: 15
- Missing Values: 0
- Duplicate Rows: 0

### 02_nav_history.csv
- Rows: 46,000
- Columns: 3
- Missing Values: 0
- Duplicate Rows: 0

### Remaining Datasets
- Most datasets contain no missing values.
- No duplicate records were detected.
- 04_monthly_sip_inflows.csv contains 12 missing values that appear to be associated with Year-over-Year growth calculations and are considered expected.

## Key Findings

- Dataset structure is consistent.
- AMFI code can be used as a reliable primary identifier.
- All AMFI codes were successfully validated against NAV history.
- No referential integrity issues were detected.
- Date columns are currently stored as strings and should be converted to datetime format during data cleaning.

## Overall Status

PASS

The datasets are suitable for further cleaning, exploratory data analysis, performance analytics and dashboard development.